tpm2/tpm2.go

```diff
@@ -1235,30 +1235,35 @@ func NVDefineSpaceEx(rw io.ReadWriter, owner tpmutil.Handle, ownerAuth, authVal
 	return err
 }
 
-func encodeWriteNV(owner, handle tpmutil.Handle, authString string, data tpmutil.U16Bytes, offset uint16) ([]byte, error) {
-	auth, err := encodeAuthArea(AuthCommand{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(authString)})
+// NVWrite writes data into the TPM's NV storage.
+func NVWrite(rw io.ReadWriter, owner, handle tpmutil.Handle, authString string, data tpmutil.U16Bytes, offset uint16) error {
+	auth := AuthCommand{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(authString)}
+	return NVWriteEx(rw, owner, handle, auth, data, offset)
+}
+
+// NVWriteEx does the same as NVWrite with the exception of letting the user take care of the AuthCommand before calling the function.
+// This allows more flexibility and does not limit the AuthCommand to PasswordSession.
+func NVWriteEx(rw io.ReadWriter, owner, index tpmutil.Handle, authArea AuthCommand, data tpmutil.U16Bytes, offset uint16) error {
+	h, err := tpmutil.Pack(owner, index)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	out, err := tpmutil.Pack(owner, handle)
+	authEnc, err := encodeAuthArea(authArea)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	buf, err := tpmutil.Pack(data, offset)
+
+	d, err := tpmutil.Pack(data, offset)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return concat(out, auth, buf)
-}
 
-// NVWrite writes data into the TPM's NV storage.
-func NVWrite(rw io.ReadWriter, owner, handle tpmutil.Handle, authString string, data tpmutil.U16Bytes, offset uint16) error {
-	Cmd, err := encodeWriteNV(owner, handle, authString, data, offset)
+	b, err := concat(h, authEnc, d)
 	if err != nil {
 		return err
 	}
-	_, err = runCommand(rw, TagSessions, CmdWriteNV, tpmutil.RawBytes(Cmd))
-	return err
+	_, err = runCommand(rw, TagSessions, CmdWriteNV, tpmutil.RawBytes(b))
+	return nil
```

Comment:
Should this line have `return err` instead? I think this is what is causing the test to fail - your code is doing the right thing but accidentally leaving the final error value behind.